### PR TITLE
added: support for adding labels from the config file itself

### LINKED ISSUE #1574: Support node labels within config files
**What would you like to be added**:

It would be nice if we had a simple:
```
- role: worker
  labels:
     type: zookeeper
     node-role.kubernetes.io/zookeeper=true: true
```

**Why is this needed**:

Using node-labels within the worker KubeAdm JoinConfiguration is:
a) Cumbersome for new users
b) Fails if you try to set a node-role which can't be specified in the JoinConfiguration

### PR DESCRIPTION
**Pull request by yashvardhan-kukreja**
closes #1574 

**My approach:**
So, the labelling of the nodes is going to be one of the actions (`actionsToRun`) when cluster creation happens in the code.
Now, I have not coded a dedicated action for "just the labelling of nodes". 
I have rather created a file `postcreationconfiguration.go` which will consist of all the functions which are supposed to occur after the cluster gets created. (like labelling in our case)

And in the file `postcreationconfiguration.go`, I have written a `labelNodes` function which will label the nodes from the config file.

**Why did I do it?**
Because, say, in the future, there is another requirement like adding taints or tolerations (_or some other configuration, which usually in real life are done after cluster is created_) through the config file itself. 
Then, we can write functions for those things in the `postcreationconfigurations.go` itself instead of creating dedicated actions for them. That's because these things like adding taints or tolerations are usually done after the cluster is already created.

**What's happening behind the scenes?**
So, basically, at every iteration, if the labels are being applied to a control-plane node, then, that control-plane-node itself will self-apply those labels to itself.
But if the labels are being applied to a worker node, then, one of the control plane nodes will apply those labels to that worker node.

**Usage from user's side**:
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: control-plane
  labels:
    name: master
    index: 2
- role: worker
  labels:
    name: worker
    tier: app
- role: worker
```

**Output**: (Look over the end of every labels row)
```
NAME                   STATUS  ROLES   LABELS
master-control-plane   Ready   master  beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,index=2,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-control-plane,kubernetes.io/os=linux,name=master,node-role.kubernetes.io/master=
master-control-plane2  Ready   master  beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-control-plane2,kubernetes.io/os=linux,node-role.kubernetes.io/master=
master-worker          Ready   <none>  beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-worker,kubernetes.io/os=linux
master-worker2         Ready   <none>  beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-worker2,kubernetes.io/os=linux,name=worker,tier=app
```

### CODE REVIEW

**Comment by BenTheElder** [comment]
Why wouldn't we have kubelet apply these?

**Comment by yashvardhan-kukreja** [comment]
> Why wouldn't we have kubelet apply these?

then, how would we go on applying labels to control plane nodes?

Are you suggesting applying labels to "worker" nodes using kubelet?

**Comment by BenTheElder** [comment]
> If we go with kubelet's way of applying node labels, then, we'd have to code different separate logics for adding labels to worker nodes (using kubelet at the time of registering nodes) and adding labels to control-plane nodes (using kubectl or something else).

Huh? All of the nodes run kubelet and are registered. The labels are attached to the node at registration time by kubelet. The control-plane kubelet registers as well ...

> Also, here, in the code, the labels are being created as a post-cluster-creation step where all the control-plane and worker nodes are being labelled together with the same function/logic. Hence, kubectl seemed the cleaner and more straightforward approach to implement the label creation.

kind is not a general command runner and we already have a (bad) abstraction for various stages of doing things.

**Comment by yashvardhan-kukreja** [comment]
> > If we go with kubelet's way of applying node labels, then, we'd have to code different separate logics for adding labels to worker nodes (using kubelet at the time of registering nodes) and adding labels to control-plane nodes (using kubectl or something else).
> 
> Huh? All of the nodes run kubelet and are registered. The labels are attached to the node at registration time by kubelet. The control-plane kubelet registers as well ...
> 
> > Also, here, in the code, the labels are being created as a post-cluster-creation step where all the control-plane and worker nodes are being labelled together with the same function/logic. Hence, kubectl seemed the cleaner and more straightforward approach to implement the label creation.
> 
> kind is not a general command runner and we already have a (bad) abstraction for various stages of doing things.

Ohh alright. My bad. Will hop into implementing with kubelet.

**Comment by yashvardhan-kukreja** [comment]
@BenTheElder , just a small thought. 

The cluster creation is happening through kubeadm in the code. And that's why , the code is not executing any kubelet commands directly while creating the cluster. (kubeadm cli is doing that, behind the scenes).
But the way you suggested; "adding labels to nodes at the time of node creation itself via kubelet" (using the `kubelet --node-labels` ). That would require the code to execute `kubelet` commands instead. 

Wouldn't that be messy or at least rough to implement? 

Just a thought. Please correct me if I am wrong.

**Comment by yashvardhan-kukreja** [comment]
Also, I don't know if there's anyway to do labelling of nodes via kubeadm but still, say, if there's a way to do so. Then, we would have to code the labelling of nodes at two places. One in the  `kubeadminit.go` action (for dealing with the labelling of control-plane nodes). and other in the `kubeadmjoin.go` action (for dealing with the labelling of worker nodes.)

But in my approach in the current PR, it's all happening in one place. 
Any thoughts regarding that?

**Comment by yashvardhan-kukreja** [comment]
I hope that makes sense. 
(please correct me if you think I am getting wrong somewhere)

**Comment by BenTheElder** [comment]
> Also, I don't know if there's anyway to do labelling of nodes via kubeadm but still, say, if there's a way to do so. Then, we would have to code the labelling of nodes at two places. One in the kubeadminit.go action (for dealing with the labelling of control-plane nodes). and other in the kubeadmjoin.go action (for dealing with the labelling of worker nodes.)

No, it would be done via the kubeadm config, by adjusting kubelet options.

File: pkg/cluster/internal/create/actions/postcreationconfiguration/postcreationconfiguration.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package postcreationconfiguration implements the action for configuring cluster after cluster creation
+// like labelling, tainting, etc the nodes.
```
Comment (BenTheElder) [inline pkg/cluster/internal/create/actions/postcreationconfiguration/postcreationconfiguration.go:18]: I'm going to delete the whole bad action abstraction when I get some uninterrupted development time anyhow, but until then, these should be scoped to specific tasks so we can easily skip specific tasks.

**Comment by yashvardhan-kukreja** [comment]
@BenTheElder I removed all that 'postcreationconfiguration' stuff and coded the addition of labels at the time of node registration via kubeadm's kubeletExtraArgs.

**Comment by yashvardhan-kukreja** [comment]
And I tested it with the following configuration and it's working perfectly fine:
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  labels:
    name: sample-controller
    age: 2
- role: control-plane
  labels:
    name: sample-controller-2
    index: 3
- role: worker
  labels:
    name: sample-worker
    worker-age: 3
- role: worker
- role: worker
```

**Output**
```
► kubectl get nodes --show-labels
NAME                  STATUS   ROLES    AGE   VERSION   LABELS
kind-control-plane    Ready    master   25m   v1.19.3   age=2,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=kind-control-plane,kubernetes.io/os=linux,name=sample-controller,node-role.kubernetes.io/master=
kind-control-plane2   Ready    master   25m   v1.19.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,index=3,kubernetes.io/arch=amd64,kubernetes.io/hostname=kind-control-plane2,kubernetes.io/os=linux,name=sample-controller-2,node-role.kubernetes.io/master=
kind-worker           Ready    <none>   23m   v1.19.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=kind-worker,kubernetes.io/os=linux
kind-worker2          Ready    <none>   23m   v1.19.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=kind-worker2,kubernetes.io/os=linux,name=sample-worker,worker-age=3
kind-worker3          Ready    <none>   23m   v1.19.3   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=kind-worker3,kubernetes.io/os=linux
```

**Comment by yashvardhan-kukreja** [comment]
@BenTheElder can you please review this?

**Comment by yashvardhan-kukreja** [comment]
@BenTheElder , sorry for pestering. Can you please review this PR?

**Comment by yashvardhan-kukreja** [comment]
@BenTheElder @neolit123 @krzyzacy Please review this PR

**Comment by BenTheElder** [comment]
sorry all time is currently consumed on bug-fixes (or non kind related work, none of the maintainers work full time on kind), with some small bits of time granted to documentation (which helps save everyone's time).

features and releases are on hold, in particular #1969 needs to be resolved.

I will get to this ASAP, but I need to prioritize these still.

**Comment by yashvardhan-kukreja** [comment]
> features and releases are on hold, in particular #1969 needs to be resolved.
Regarding this, this has already received an lgtm from Lubomir :)

> sorry all time is currently consumed on bug-fixes (or non kind related work, none of the maintainers work full time on kind), with some small bits of time granted to documentation (which helps save everyone's time).
> 
> features and releases are on hold, in particular #1969 needs to be resolved.
> 
> I will get to this ASAP, but I need to prioritize these still.
> 
> EDIT: previously I was personally unavailable for a bit for any work for personal reasons, since I've been back I've been resolving those.

Sure @BenTheElder . No worries. Please take your time.
Thanks a lot for the heads up though 😄

**Comment by BenTheElder** [comment]
https://github.com/kubernetes-sigs/kind/issues/2023

**Comment by BenTheElder** [comment]
sorry for the incredibly long delay. I am closing out our backlog prior to #2120 and since the last release we've expanded the maintainer group to help avoid this.

I'd still like to move this to v1alpha5, but as that doesn't exist yet, merging as v1alpha4. if we get v1alpha5 ready in time for 1.21, we'll migrate this over. if not we'll just leave it as a v1alpha4 addition.

thank you for working on this.

**Comment by wallrj** [comment]
Some node labels in the config file (such as node-role.kubernetes.io/worker)  cause `kind create cluster` to fail with an unfriendly error

I attempted to use this feature to set a meaningful role name for the worker nodes.
Here's a demonstration of manually labelling the nodes and the effect it has on the output of `kubectl get nodes`

```
$ kubectl get nodes
NAME                 STATUS   ROLES           AGE   VERSION
kind-control-plane   Ready    control-plane   73s   v1.27.3
kind-worker          Ready    <none>          49s   v1.27.3
kind-worker2         Ready    <none>          49s   v1.27.3

$ kubectl label node kind-worker node-role.kubernetes.io/platform=
node/kind-worker labeled

$ kubectl get nodes
NAME                 STATUS   ROLES           AGE   VERSION
kind-control-plane   Ready    control-plane   97s   v1.27.3
kind-worker          Ready    platform        73s   v1.27.3
kind-worker2         Ready    <none>          73s   v1.27.3
```

But with the following config file, `kind create cluster` just crashes after a number of minutes.

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
  labels:
    node-role.kubernetes.io/worker: ""
```

```sh
$ kind create cluster --config kind.config.yaml
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.29.2) 🖼
 ✓ Preparing nodes 📦 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
 ✗ Joining worker nodes 🚜
Deleted nodes: ["kind-control-plane" "kind-worker"]
ERROR: failed to create cluster: failed to join node with kubeadm: command "docker exec --privileged kind-worker kubeadm join --config /kind/kubeadm.conf --skip-phases=preflight --v=6" failed with error: exit status 1
Command Output: I0228 13:43:06.510544     140 join.go:413] [preflight] found NodeName empty; using OS hostname as NodeName
I0228 13:43:06.510617     140 joinconfiguration.go:76] loading configuration from "/kind/kubeadm.conf"
I0228 13:43:06.511206     140 controlplaneprepare.go:225] [download-certs] Skipping certs download
I0228 13:43:06.511234     140 join.go:532] [preflight] Discovering cluster-info
I0228 13:43:06.511247     140 token.go:80] [discovery] Created cluster-info discovery client, requesting info from "kind-control-plane:6443"
I0228 13:43:06.517396     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/namespaces/kube-public/configmaps/cluster-info?timeout=10s 200 OK in 5 milliseconds
I0228 13:43:06.517665     140 token.go:223] [discovery] The cluster-info ConfigMap does not yet contain a JWS signature for token ID "abcdef", will try again
I0228 13:43:12.610900     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/namespaces/kube-public/configmaps/cluster-info?timeout=10s 200 OK in 5 milliseconds
I0228 13:43:12.611048     140 token.go:223] [discovery] The cluster-info ConfigMap does not yet contain a JWS signature for token ID "abcdef", will try again
I0228 13:43:18.173233     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/namespaces/kube-public/configmaps/cluster-info?timeout=10s 200 OK in 4 milliseconds
I0228 13:43:18.173408     140 token.go:223] [discovery] The cluster-info ConfigMap does not yet contain a JWS signature for token ID "abcdef", will try again
I0228 13:43:23.728172     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/namespaces/kube-public/configmaps/cluster-info?timeout=10s 200 OK in 7 milliseconds
I0228 13:43:23.729774     140 token.go:105] [discovery] Cluster info signature and contents are valid and no TLS pinning was specified, will use API Server "kind-control-plane:6443"
I0228 13:43:23.729817     140 discovery.go:52] [discovery] Using provided TLSBootstrapToken as authentication credentials for the join process
I0228 13:43:23.729831     140 join.go:546] [preflight] Fetching init configuration
I0228 13:43:23.729836     140 join.go:592] [preflight] Retrieving KubeConfig objects
[preflight] Reading configuration from the cluster...
[preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
I0228 13:43:23.739778     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config?timeout=10s 200 OK in 9 milliseconds
I0228 13:43:23.740952     140 kubeproxy.go:55] attempting to download the KubeProxyConfiguration from ConfigMap "kube-proxy"
I0228 13:43:23.744918     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/namespaces/kube-system/configmaps/kube-proxy?timeout=10s 200 OK in 3 milliseconds
I0228 13:43:23.746979     140 kubelet.go:74] attempting to download the KubeletConfiguration from ConfigMap "kubelet-config"
I0228 13:43:23.750225     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/namespaces/kube-system/configmaps/kubelet-config?timeout=10s 200 OK in 2 milliseconds
I0228 13:43:23.756648     140 initconfiguration.go:114] skip CRI socket detection, fill with the default CRI socket unix:///var/run/containerd/containerd.sock
I0228 13:43:23.756874     140 interface.go:432] Looking for default routes with IPv4 addresses
I0228 13:43:23.756911     140 interface.go:437] Default route transits interface "eth0"
I0228 13:43:23.756998     140 interface.go:209] Interface eth0 is up
I0228 13:43:23.757038     140 interface.go:257] Interface "eth0" has 3 addresses :[172.18.0.2/16 fc00:f853:ccd:e793::2/64 fe80::42:acff:fe12:2/64].
I0228 13:43:23.757055     140 interface.go:224] Checking addr  172.18.0.2/16.
I0228 13:43:23.757066     140 interface.go:231] IP found 172.18.0.2
I0228 13:43:23.757076     140 interface.go:263] Found valid IPv4 address 172.18.0.2 for interface "eth0".
I0228 13:43:23.757085     140 interface.go:443] Found active IP 172.18.0.2
I0228 13:43:23.763621     140 kubelet.go:121] [kubelet-start] writing bootstrap kubelet config file at /etc/kubernetes/bootstrap-kubelet.conf
I0228 13:43:23.764767     140 kubelet.go:136] [kubelet-start] writing CA certificate at /etc/kubernetes/pki/ca.crt
I0228 13:43:23.765328     140 loader.go:395] Config loaded from file:  /etc/kubernetes/bootstrap-kubelet.conf
I0228 13:43:23.765713     140 kubelet.go:157] [kubelet-start] Checking for an existing Node in the cluster with name "kind-worker" and status "Ready"
I0228 13:43:23.768761     140 round_trippers.go:553] GET https://kind-control-plane:6443/api/v1/nodes/kind-worker?timeout=10s 404 Not Found in 2 milliseconds
I0228 13:43:23.769169     140 kubelet.go:172] [kubelet-start] Stopping the kubelet
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[kubelet-start] Starting the kubelet
[kubelet-start] Waiting for the kubelet to perform the TLS Bootstrap...
[kubelet-check] Initial timeout of 40s passed.
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.

Unfortunately, an error has occurred:
        timed out waiting for the condition

This error is likely caused by:
        - The kubelet is not running
        - The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)

If you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:
        - 'systemctl status kubelet'
        - 'journalctl -xeu kubelet'
timed out waiting for the condition
error execution phase kubelet-start
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
        cmd/kubeadm/app/cmd/phases/workflow/runner.go:260
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
        cmd/kubeadm/app/cmd/phases/workflow/runner.go:446
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
        cmd/kubeadm/app/cmd/phases/workflow/runner.go:232
k8s.io/kubernetes/cmd/kubeadm/app/cmd.newCmdJoin.func1
        cmd/kubeadm/app/cmd/join.go:180
github.com/spf13/cobra.(*Command).execute
        vendor/github.com/spf13/cobra/command.go:940
github.com/spf13/cobra.(*Command).ExecuteC
        vendor/github.com/spf13/cobra/command.go:1068
github.com/spf13/cobra.(*Command).Execute
        vendor/github.com/spf13/cobra/command.go:992
k8s.io/kubernetes/cmd/kubeadm/app.Run
        cmd/kubeadm/app/kubeadm.go:50
main.main
        cmd/kubeadm/kubeadm.go:25
runtime.main
        /usr/local/go/src/runtime/proc.go:267
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1650
```

**Pull request merged**